Memory leak risk: Fix 'cancelTransformableRequest' function when using 'remove_if' caus…

straybird-s opened the pull request:
I used TF2 in my own project. When I do '**waitForTransform**' CI test, I found a memory leak and finally focused the problem at 'cancelTransformableRequest' function: 
If vector 'transformable_requests_' has more than 1 elements, 'transformable_callbacks_.erase()' can't work correctly.  That will make **'transformable_callbacks_.size()' rising**. 
The reason for this issue is 'remove' can cause unspecified values. The details refer to: ( https://en.cppreference.com/w/cpp/algorithm/remove )
"Iterators pointing to an element between the new logical end and the physical end of the range are still dereferenceable, **but the elements themselves have unspecified values**"

straybird-s committed:
Fix 'cancelTransformableRequest' function when using 'remove_if' cause unspecified values.

aprotyas commented:
Correct me if I'm wrong, but the problem here is that `it->cb_handle` is some unspecified value, and not the actual `cb_handle` corresponding to that iterator. Is it?

https://github.com/ros2/geometry2/blob/16562cee8694c11f1f82b2bdbde2814fca1c7954/tf2/src/buffer_core.cpp#L1212-L1216

That makes sense to me because things past the logical end have been moved there, so they shouldn't actually be dereferenced.

aprotyas commented on tf2/src/buffer_core.cpp:
```diff
@@ -1209,13 +1209,18 @@ void BufferCore::cancelTransformableRequest(TransformableRequestHandle handle)
+  
+  if (it != transformable_requests_.end()) {
     transformable_callbacks_.erase(it->cb_handle);
+    if (transformable_requests_.size() > 1) {
+      transformable_requests_[it -
+        transformable_requests_.begin()] = transformable_requests_.back();
```
Can you explain what's happening here?

Imagine `transformable_requests_.size() == 5`, and `it` points to the first element, i.e. `it == transformable_requests_.begin()`. Then, once this section is executed, you'd have `transformable_requests_[0] = transformable_requests_.back()`? What would that achieve?

That logic doesn't checkout for me.

straybird-s replied:
The solution that I used here is referenced from the 'BufferCore::testTransformableRequests()' function:
https://github.com/ros2/geometry2/blob/16562cee8694c11f1f82b2bdbde2814fca1c7954/tf2/src/buffer_core.cpp#L1305-L1310
The idea is the same: replace the last value to the location on where the value to be deleted, than erase the last element that has been preserved in front.

aprotyas replied:
I think I follow what's happening here for the most part, maybe I was wrong in the assumption that there could be multiple requests with the same handle.

straybird-s replied:
Certainly, considering the non-unique case increases robustness. But I want to say that the 'addTransformableRequest' function guarantees uniqueness of the 'request_handle' in 'transformable_requests_' vector, unless the sum of the TransformableRequest more than 0xffffffffffffffffULL. If that extreme conditions occur, one handle ID refer to different callbacks, that may cause troubles. So I think non-unique handles cases will not be allowed.

aprotyas commented:
After having thought about the problem here, I think this patch better represents what should be happening. What do you think @straybird-s?

```diff
<   V_TransformableRequest::iterator remove_it = std::remove_if(transformable_requests_.begin(), transformable_requests_.end(),
<                                                               [handle](TransformableRequest req) { return handle == req.request_handle; });
<   for (V_TransformableRequest::iterator it = remove_it; it != transformable_requests_.end(); ++it) {
<     transformable_callbacks_.erase(it->cb_handle);
<   }
---
>   auto remove_it = std::remove_if(
>     transformable_requests_.begin(),
>     transformable_requests_.end(),
>     [handle](const auto & req) {
>       if (handle == req.request_handle) {
>         transformable_callbacks_.erase(req.cb_handle);
>         return true;
>       }
>       return false;
>     });
```

Having a predicate with side effects - modifying the `transformable_callbacks_` map in the lambda - is not ideal, but the alternatives are doing a separate pre-processing scan or rolling our own remove_if-erase idiom which relaxes the movable constraint you mentioned.

---

[Edit]: The "predicate-with-side-effects" was bothering me, and I thought of this alternative which captures the semantics better, I think. Thoughts? I'm not sure if the relative ordering of requests are important, but if so `std::partition` can be swapped out for `std::stable_partition`.

```cpp
  // push requests with matching handles to the end
  auto remove_it = std::partition(
    transformable_requests_.begin(),
    transformable_requests_.end(),
    [handle](const auto & req) { return handle != req.request_handle; });

  // remove callbacks from map
  std::for_each(
    remove_it,
    transformable_requests_.end(),
    [](const auto & request) {
      transformable_callbacks_.erase(request->cb_handle);
    });

  // remove matching requests altogether
  transformable_requests_.erase(remove_it, transformable_requests_.end());
```

straybird-s commented:
Yeah, I agree that modern c++ style code is better. @aprotyas

straybird-s committed:
Robust to cancel transformable request.

aprotyas reviewed (approved):
Changes LGTM, but can you please post some more data/information about the failed test and/or the memory leak you observed for the sake of other reviewers? That will help the package maintainers better understand the problem and assess the need/appropriateness of the change.

aprotyas commented:
Running CI.
Repos file: https://gist.github.com/aprotyas/069a58bb7dcf70c2bc9aa7a7cf04e66a
Build args: `--packages-above-and-dependencies tf2`
Test args: `--packages-above tf2`
Job: https://ci.ros2.org/job/ci_launcher/9347/
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=15613)](http://ci.ros2.org/job/ci_linux/15613/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=10294)](http://ci.ros2.org/job/ci_linux-aarch64/10294/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=13281)](http://ci.ros2.org/job/ci_osx/13281/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=15829)](http://ci.ros2.org/job/ci_windows/15829/)

straybird-s commented:
The test as follows:

add `print` of  `transformable_callbacks_.size()` in `geometry2/tf2/src/buffer_core.cpp`:

```cpp
   V_TransformableRequest::iterator remove_it = std::remove_if(transformable_requests_.begin(), transformable_requests_.end(),
                                                               [handle](TransformableRequest req) { return handle == req.request_handle; });
   for (V_TransformableRequest::iterator it = remove_it; it != transformable_requests_.end(); ++it) {
     transformable_callbacks_.erase(it->cb_handle);
     printf("The map size of transformable_callbacks: %zu\n", transformable_callbacks_.size());
   }
```
and run `test_waitForTransform.cpp`. The `print` will show increasing growth of `transformable_callbacks_` in test.

```cpp
#include <tf2_ros/buffer.h>
#include <tf2_ros/transform_listener.h>
#include <tf2_ros/transform_broadcaster.h>
#include <tf2_ros/static_transform_broadcaster.h>
#include <tf2_ros/create_timer_ros.h>
#include <rclcpp/rclcpp.hpp>

#include <thread>
#include <iostream>
#include <iomanip>
#include <sstream>

class Object {
  public:
  explicit Object(rclcpp::Node::SharedPtr & node) {
    node_ = node;
    buffer_ = std::make_shared<tf2_ros::Buffer>(node->get_clock());
    tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*buffer_, node, false);
    tf_broadcaster_ = std::make_shared<tf2_ros::TransformBroadcaster>(node);
    static_broadcaster_ = std::make_shared<tf2_ros::StaticTransformBroadcaster>(node);

    // timer
    auto create_timer_interface = std::make_shared<tf2_ros::CreateTimerROS>(
      node->get_node_base_interface(),
      node->get_node_timers_interface());
    buffer_->setCreateTimerInterface(create_timer_interface);

    geometry_msgs::msg::TransformStamped tf;
    // base_link -> camera_link
    tf.header.frame_id = "base_link";
    tf.header.stamp = node->get_clock()->now();
    tf.child_frame_id = "camera_link";
    tf.transform.rotation.x = 0.; 
    tf.transform.rotation.y = 0.; 
    tf.transform.rotation.z = 0.;
    tf.transform.rotation.w = 1.; 
    tf.transform.translation.x = 0.; 
    tf.transform.translation.y = 0.; 
    tf.transform.translation.z = 1.;
    static_broadcaster_->sendTransform(tf);

    pub_thread_ = new std::thread([this, tf]() {
      float step = 1.f;
      geometry_msgs::msg::TransformStamped transform = tf;
      while (1) {
        transform.header.frame_id = "odom";
        transform.header.stamp = node_->get_clock()->now();
        transform.child_frame_id = "base_link";
        transform.transform.translation.z += step;
        tf_broadcaster_->sendTransform(transform);
        std::this_thread::sleep_for(std::chrono::milliseconds(50));
      }
    });

    pub_thread2_ = new std::thread([this, tf]() {
      float step = -1.f;
      geometry_msgs::msg::TransformStamped transform = tf;
      while (1) {
        transform.header.frame_id = "map";
        transform.header.stamp = node_->get_clock()->now();
        transform.child_frame_id = "odom";
        transform.transform.translation.z += step;
        tf_broadcaster_->sendTransform(transform);
        std::this_thread::sleep_for(std::chrono::milliseconds(50));
      }
    });

    listen_thread_ = new std::thread([this]() {
      geometry_msgs::msg::TransformStamped transform;
      std::ostringstream oss;
      while (1) {
        oss.clear();
        auto future = buffer_->waitForTransform(
          "base_link",
          "odom",
          node_->get_clock()->now(), tf2::durationFromSec(0.01),
          [&](const tf2_ros::TransformStampedFuture & future)
          {
            try
            {
              transform = future.get();
            }
            catch(const std::exception& e)
            {
              transform = buffer_->lookupTransform(
                "base_link", "odom", tf2::timeFromSec(0));
            }
            oss << std::fixed << std::setprecision(4) 
              << tf2_ros::timeToSec(transform.header.stamp) << ": "
              << transform.header.frame_id << " -> "
              << transform.child_frame_id << " | "
              << "z: " << transform.transform.translation.z << std::endl;
          });
        std::this_thread::sleep_for(std::chrono::milliseconds(50));
        std::cout << oss.str();
      }
    });

    listen_thread2_ = new std::thread([this]() {
      geometry_msgs::msg::TransformStamped transform;
      std::ostringstream oss;
      while (1) {
        oss.clear();
        auto future = buffer_->waitForTransform(
          "odom",
          "map",
          node_->get_clock()->now(), tf2::durationFromSec(0.01),
          [&](const tf2_ros::TransformStampedFuture & future)
          {
            try
            {
              transform = future.get();
            }
            catch(const std::exception& e)
            {
              transform = buffer_->lookupTransform(
                "odom", "map", tf2::timeFromSec(0));
            }
            oss << std::fixed << std::setprecision(4) 
              << tf2_ros::timeToSec(transform.header.stamp) << ": "
              << transform.header.frame_id << " -> "
              << transform.child_frame_id << " | "
              << "z: " << transform.transform.translation.z << std::endl;
          });
        std::this_thread::sleep_for(std::chrono::milliseconds(50));
        std::cout << oss.str();
      }
    });
  }

  Object() = delete;

  ~Object() {}

  private:
  rclcpp::Node::SharedPtr node_;
  std::shared_ptr<tf2_ros::Buffer> buffer_;
  std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
  std::shared_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;
  std::shared_ptr<tf2_ros::StaticTransformBroadcaster> static_broadcaster_;

  //Thread
  std::thread* pub_thread_;
  std::thread* pub_thread2_;
  std::thread* listen_thread_;
  std::thread* listen_thread2_;
};

int main(int argc, char** argv) 
{
  rclcpp::init(argc, argv);

  auto node = rclcpp::Node::make_shared("tf2_ros_waitForTransform");
  Object object(node);

  rclcpp::spin(node);
}
```

aprotyas commented:
@straybird-s is the memory leak no longer happening? If not, since there's a ROS 2 org member assigned to this ticket, I don't think this PR should be closed for now.

straybird-s committed:
Merge branch 'ros2' into tf2_debug

cyan9527 commented:
I encountered this memory leak issue in Humble, but the PR has been closed. Has it been fixed? @aprotyas

aprotyas commented:
I don't think this has been fixed. The PR never merged and I still see the wrong code on the rolling branch. @straybird-s do you want to re-open this? Pinging @clalancette if you want to help merge this memory leak fix!

roehling commented:
This PR really should be reopened, because it is not just a memory leak. The invalid callback handles might actually refer to an unrelated callback, which gets erased by mistake. The following program demonstrates the problem:

```c++
#include <rclcpp/time.hpp>
#include <tf2/buffer_core.hpp>

#include <iostream>

void callback(tf2::TransformableRequestHandle request_handle, const std::string& target_frame,
              const std::string& source_frame, const tf2::TimePoint& time, tf2::TransformableResult result)
{
    std::cout << "callback(request_handle=" << request_handle << ", target_frame=\"" << target_frame
              << "\", source_frame=\"" << source_frame << "\", time=" << tf2::displayTimePoint(time)
              << ", result=" << result << ")\n";
}

template<typename Rep, typename Period>
tf2::TransformableRequestHandle add_transformable_request(tf2::BufferCore& bc, const std::string& target_frame,
                                                          const std::string& source_frame,
                                                          const std::chrono::duration<Rep, Period>& time)
{
    tf2::TimePoint tp{time};
    tf2::TransformableRequestHandle h = bc.addTransformableRequest(callback, target_frame, source_frame, tp);
    std::cout << "add_transformable_request(target_frame=\"" << target_frame << "\", source_frame=\"" << source_frame
              << "\", time=" << tf2::displayTimePoint(tp) << ") -> " << h << "\n";
    return h;
}

template<typename Rep, typename Period>
void set_transform(tf2::BufferCore& bc, const std::string& target_frame, const std::string& source_frame,
                   const std::chrono::duration<Rep, Period>& time)
{
    std::chrono::nanoseconds ns = std::chrono::duration_cast<std::chrono::nanoseconds>(time);
    geometry_msgs::msg::TransformStamped tr;
    tr.header.frame_id = target_frame;
    tr.header.stamp = rclcpp::Time(ns.count());
    tr.child_frame_id = source_frame;
    tr.transform.rotation.w = 1;  // make transform valid
    std::cout << "set_transform(target_farme=\"" << target_frame << "\", source_frame=\"" << source_frame
              << "\", time=" << tf2::displayTimePoint(tf2::TimePoint(time)) << ")\n";
    bc.setTransform(tr, "test_injector", false);
}

void cancel_transformable_request(tf2::BufferCore& bc, tf2::TransformableRequestHandle request_handle)
{
    std::cout << "cancel_transformable_request(request_handle=" << request_handle << ")\n";
    bc.cancelTransformableRequest(request_handle);
}

int main()
{
    using namespace std::chrono_literals;
    tf2::BufferCore bc{10s};
    tf2::TransformableRequestHandle handle = add_transformable_request(bc, "target", "alpha", 100s);
    add_transformable_request(bc, "target", "beta", 100s);
    add_transformable_request(bc, "target", "gamma", 100s);
    cancel_transformable_request(bc, handle);
    set_transform(bc, "target", "alpha", 100s);
    set_transform(bc, "target", "beta", 100s);
    set_transform(bc, "target", "gamma", 100s);
    return 0;
}
```

The program output is as follows:
```
add_transformable_request(target_frame="target", source_frame="alpha", time=100.000000) -> 1
add_transformable_request(target_frame="target", source_frame="beta", time=100.000000) -> 2
add_transformable_request(target_frame="target", source_frame="gamma", time=100.000000) -> 3
cancel_transformable_request(request_handle=1)
set_transform(target_farme="target", source_frame="alpha", time=100.000000)
set_transform(target_farme="target", source_frame="beta", time=100.000000)
callback(request_handle=2, target_frame="target", source_frame="beta", time=100.000000, result=0)
set_transform(target_farme="target", source_frame="gamma", time=100.000000)
```

As you can see, the cancellation of the `alpha` request inadvertently also destroyed the callback for the `gamma` request.
